src/Storages/StorageMergeTree.cpp

```diff
@@ -1428,72 +1409,288 @@ MergeTreeDataPartPtr StorageMergeTree::outdatePart(MergeTreeTransaction * txn, c
 void StorageMergeTree::dropPartNoWaitNoThrow(const String & part_name)
 {
     if (auto part = outdatePart(NO_TRANSACTION_RAW, part_name, /*force=*/ false))
-        dropPartsImpl({part}, /*detach=*/ false);
+    {
+        if (deduplication_log)
+        {
+            deduplication_log->dropPart(part->info);
+        }
+
+        LOG_INFO(log, "Removed 1 parts.");
+
+        /// Need to destroy part objects before clearing them from filesystem.
+        part.reset();
+
+        clearOldPartsFromFilesystem();
+    }
 
     /// Else nothing to do, part was removed in some different way
 }
 
-void StorageMergeTree::dropPart(const String & part_name, bool detach, ContextPtr query_context)
+using RangesWithContinuousBlocks = std::vector<DataPartsVector>;
+
+RangesWithContinuousBlocks groupByRangesWithContinuousBlocks(DataPartsVector parts)
 {
-    if (auto part = outdatePart(query_context->getCurrentTransaction().get(), part_name, /*force=*/ true))
-        dropPartsImpl({part}, detach);
+    RangesWithContinuousBlocks result;
+
+    std::sort(parts.begin(), parts.end(), MergeTreeData::LessDataPart());
+
+    for (const auto & part: parts)
+    {
+        if (result.empty())
+        {
+            result.push_back({part});
+            continue;
+        }
+
+        auto last_part_in_prev_range = result.back().back();
+        if (last_part_in_prev_range->info.max_block+1 == part->info.min_block)
```

Comment:
Seems like for StorageMergeTree this condition will be almost always false if user inserts data into multiple partitions:
```
dell9510 :) create table t (n int) engine=MergeTree order by n partition by n % 10
dell9510 :) insert into t values (11)
dell9510 :) insert into t values (22)
dell9510 :) insert into t values (31)
dell9510 :) insert into t values (42)
dell9510 :) insert into t values (51)

dell9510 :) select *, _part from t order by _part

┌──n─┬─_part───┐
│ 11 │ 1_1_1_0 │
└────┴─────────┘
┌──n─┬─_part───┐
│ 31 │ 1_3_3_0 │
└────┴─────────┘
┌──n─┬─_part───┐
│ 51 │ 1_5_5_0 │
└────┴─────────┘
┌──n─┬─_part───┐
│ 22 │ 2_2_2_0 │
└────┴─────────┘
┌──n─┬─_part───┐
│ 42 │ 2_4_4_0 │
└────┴─────────┘
```

Reply:
Yes it is ring. It combine the parts only if it could. In your example new_parts.size() == old_parts.size(). No profit. 
Only if user uses time as a partition or no partition there would be lesser new_parts.

Reply:
Maybe we can remove this code then and simply create `old_parts.size()` parts?

Reply:
Actually I agree with that now. That optimisation  doesn't worse the efforts.

Not only because we could get rid of that function with grouping. The function small and simple. There is also some simplification in place where different operation have a race when they commiting their parts. With grouping intersections could happen. Without grouping they just cover each other.